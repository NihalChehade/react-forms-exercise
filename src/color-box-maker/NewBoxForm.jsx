import React, { useState } from "react";

const NewBoxForm = ({ addBox }) => {
  const INITIAL_STATE = {
    color: "#000000",
    height: "",
    width: "",
  };

  const [box, setBox] = useState(INITIAL_STATE);
  const handleSubmit = (e) => {
    e.preventDefault();
    addBox(box);
    setBox(INITIAL_STATE);
  };

  const handleChange = (e) => {
    const {name, value} = e.target;
    setBox(() => ({
      ...box,
      [name]:value
    }));
  };

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="color">Choose Background Color:</label>
      <input 
      type="color" 
      name="color" 
      id="color"  
      value ={box.color}
      onChange={handleChange}
      /><br/>

      <label htmlFor="width">Choose Box width:</label>
      <input 
      type="text" 
      name="width" 
      id="width"
      value ={box.width} 
      onChange={handleChange}
      /><br/>

      <label htmlFor="height">Choose Box Height:</label>
      <input 
      type="text" 
      name="height" 
      id="height"
      value ={box.height}
      onChange={handleChange} 
      /><br/>
      
      <button type="submit">Add Box!</button>
    </form>
  );
};

export default NewBoxForm;
