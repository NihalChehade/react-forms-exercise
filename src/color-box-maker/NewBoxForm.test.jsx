import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

// Smoke Test
test('renders NewBoxForm component without crashing', () => {
  render(<NewBoxForm addBox={() => {}} />);
});

// Snapshot Test
test('matches snapshot', () => {
  const { asFragment } = render(<NewBoxForm addBox={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});

// Critical Business Logic Tests
test('updates state when inputs change', () => {
  render(<NewBoxForm addBox={() => {}} />);
  
  const colorInput = screen.getByLabelText('Choose Background Color:');
  const widthInput = screen.getByLabelText('Choose Box width:');
  const heightInput = screen.getByLabelText('Choose Box Height:');
  
  fireEvent.change(colorInput, { target: { value: '#ff0000' } });
  fireEvent.change(widthInput, { target: { value: '200' } });
  fireEvent.change(heightInput, { target: { value: '200' } });
  
  expect(colorInput.value).toBe('#ff0000');
  expect(widthInput.value).toBe('200');
  expect(heightInput.value).toBe('200');
});

test('calls addBox with correct data and resets form on submit', () => {
  const mockAddBox = vi.fn();
  render(<NewBoxForm addBox={mockAddBox} />);
  
  const colorInput = screen.getByLabelText('Choose Background Color:');
  const widthInput = screen.getByLabelText('Choose Box width:');
  const heightInput = screen.getByLabelText('Choose Box Height:');
  
  fireEvent.change(colorInput, { target: { value: '#ff0000' } });
  fireEvent.change(widthInput, { target: { value: '200' } });
  fireEvent.change(heightInput, { target: { value: '200' } });
  
  fireEvent.click(screen.getByText('Add Box!'));
  
  expect(mockAddBox).toHaveBeenCalledWith({ color: '#ff0000', width: '200', height: '200' });
  expect(colorInput.value).toBe('#000000');
  expect(widthInput.value).toBe('');
  expect(heightInput.value).toBe('');
});
