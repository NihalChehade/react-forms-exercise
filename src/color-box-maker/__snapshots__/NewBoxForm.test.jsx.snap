// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Choose Background Color:
    </label>
    <input
      id="color"
      name="color"
      type="color"
      value="#000000"
    />
    <br />
    <label
      for="width"
    >
      Choose Box width:
    </label>
    <input
      id="width"
      name="width"
      type="text"
      value=""
    />
    <br />
    <label
      for="height"
    >
      Choose Box Height:
    </label>
    <input
      id="height"
      name="height"
      type="text"
      value=""
    />
    <br />
    <button
      type="submit"
    >
      Add Box!
    </button>
  </form>
</DocumentFragment>
`;
